keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt

```diff
@@ -12,78 +12,86 @@ import com.netflix.spinnaker.keel.api.randomUID
    *
-   * @return a hydrated `Resource` with a name generated by convention, a `uid`, default values
-   * applied, etc.
+   * @return a hydrated `Resource` with a name generated by convention, an `id`,  `uid`, etc.
    */
   fun normalize(resource: SubmittedResource<S>): Resource<S> {
     val metadata = resource.metadata + mapOf(
       "id" to resource.id.toString(),
       "uid" to randomUID().toString(),
       "application" to resource.spec.application
     )
-    val hydratedResource = Resource(
+    return Resource(
       apiVersion = resource.apiVersion,
       kind = resource.kind,
       metadata = metadata,
       spec = resource.spec
     )
-    return normalize(hydratedResource)
   }
 
   /**
-   * Validates the resource and applies any defaults / normalization.
+   * Applies any defaults / opinions to the resource as it is resolved into its [desired] state.
    *
-   * @return [resource] or a copy of [resource] that may have been changed, for example in order to
-   * set default values.
+   * @return [resource] or a copy of [resource] that may have been changed in order to set default
+   * values or apply opinions.
    */
-  fun normalize(resource: Resource<S>): Resource<S> =
-    normalizers
+  private fun applyResolvers(resource: Resource<S>): Resource<S> =
+    resolvers
       .filter { it.handles(resource.apiVersion, resource.kind) }
-      .filterIsInstance<ResourceNormalizer<S>>()
-      .fold(resource) { r, normalizer ->
-        log.debug("Normalizing ${r.id} with ${normalizer.javaClass}")
-        normalizer.normalize(r)
+      .filterIsInstance<Resolver<S>>()
+      .fold(resource) { r, resolver ->
+        log.debug("Applying ${resolver.javaClass} to ${r.id}")
+        resolver(r)
       }
 
   /**
-   * Resolve the resource spec into the desired state. This may involve looking up referenced
-   * resources, etc.
+   * Resolve and convert the resource spec into the type that represents the diff-able desired
+   * state.
    *
    * The value returned by this method is used as the basis of the diff (with the result of
    * [current] in order to decide whether to call [create]/[update]/[upsert].
    *
+   * @param resource the resource as persisted in the Keel database.
+   */
+  suspend fun desired(resource: Resource<S>): R = toResolvedType(applyResolvers(resource))
+
+  /**
+   * Convert the resource spec into the type that represents the diff-able desired state. This may
+   * involve looking up referenced resources, splitting a multi-region resource into discrete
+   * objects for each region, etc.
+   *
    * Implementations of this method should not actuate any changes.
+   *
+   * @param resource a fully-resolved version of the persisted resource spec. You can assume that
+   * [applyResolvers] has already been called on this object.
    */
-  suspend fun desired(resource: Resource<S>): R
+  protected abstract suspend fun toResolvedType(resource: Resource<S>): R
```

Comment:
The only thing individual handlers need to do is convert from spec to realized type